### ISSUE TITLE
Decouple routes from content folder

### ISSUE THREAD

**Issue by dillonkearns** [dillonkearns/elm-pages#76]
I’m working on the design now for decoupling the `elm-pages` routes from the `content/` folder. My goal is to 1) have a helper that sets up `elm-pages` exactly as it is now with mapping `content/` files to routes. But, 2) give you the flexibility to create routes either based on files, StaticHttp response data… or just a hardcoded path with no associated file or static data! Or a route based on data from 2 different files!

My current thinking is that the design will look something like this.

* Give a helper that allows you to iterate over files and turn those files into `Static` data.
* Return a `CreatePage` record that represents how to pull in from data sources, decode, reduce down the JSON, and then create a route with that view and metadata

```elm
type alias CreatePage metadata pageView =
    Result String 
    { path : List String
    , metadata : metadata
    , content : pageView
    }

createPages =
Static.findFiles { extension = “md”, base = [ “content” ] }
|> Static.map
  (List.map (\file -> 
    Static.map2 (\meta body ->
      { route = file.path, metadata = meta, body = body }
    )
    (file
      |> Static.frontmatter 
      (Decode.map2 Metadata.Page
         (Decode.field “title” Decode.string)
         (Decode.field “tags” (Decode.list Decode.string))
      )
    )
    (file
      |> Static.body {- StaticHttp String -}
      |> Static.map parseMarkdown
    )

  )
)
```

**Comment by sevfurneaux** [comment]
I'm super excited for this to be implemented! I think this feature will ultimately make it parallel with Gatsby.

**Comment by dillonkearns** [comment]
Yes, I was thinking the same! I think at that point you'll be able to build most things that you could with Gatsby. I'm also working on an image optimization API, so at that point it will be comparable in what it can be used for.

I've got a few steps lined up for getting this feature out. It's going to  be a big architectural change under the hood, but really exciting stuff!

**Comment by sevfurneaux** [comment]
Awesome! Good to know what needs to be done prior to this launching 👍

**Comment by tgelu** [comment]
+100

**Comment by sevfurneaux** [comment]
I was thinking about this feature recently, and was wondering how one would generate a dynamic route which also had an `id` in the slug? 🤯

Before this feature is implemented, I was considering programmatically creating a markdown files using JavaScript _before_ the build step, similar to what @tgelu mentions in https://github.com/dillonkearns/elm-pages/issues/108.

Nonetheless, would to implement properly using this new feature approach 💪

**Comment by sevfurneaux** [comment]
I'm working on an elm-pages project where I need to programatically create the content, so I created this script which gets data from multiple endpoints and then creates the subsequent folders and saves the JSON data as frontmatter in the markdown files:

```
const fetch = require("node-fetch");
var fs = require("fs").promises;
var path = require("path");

const ENDPOINTS = [
  "http://example.com/endpoint-1",
  "http://example.com/endpoint-2",
  "http://example.com/endpoint-3",
];

ENDPOINTS.forEach(endpoint => createFilesFrom(endpoint));

function createFilesFrom(endpoint) {
  const CONTENT_PATH = "content";

  fetch(endpoint)
    .then(response => response.json())
    .then(response => {
      const data = response.data;

      if (!response.data) {
        throw new Error("No data response.");
      }

      data.forEach(page => {
        const url = page.url
          .substring(1);

        if (url === "") {
          createFile(`${CONTENT_PATH}/index.md`, page);
        } else {
          createFile(`${CONTENT_PATH}/${url}.md`, page);
        }
      });
    });
}

function createFile(filePath, content) {
  fs.mkdir(path.dirname(filePath), { recursive: true }).then(() =>
    fs.writeFile(filePath, `---\n${JSON.stringify(content, null, 2)}\n---`)
  );
}
```

Might be useful for folk working with dynamic CMS content prior to dynamic/programmatic routes being implemented.

**Comment by avh4** [comment]
> Or a route based on data from 2 different files!
I'm interested in this for the elm-format docs

**Comment by dillonkearns** [comment]
This issue is resolved now as of the 2.0.0 NPM package/8.0.0 Elm package package! 🎉
